### PR TITLE
Add the endpoints data source object

### PR DESCRIPTION
**Pull request by Dudesons**
Signed-off-by: Damien Goldenberg <damien@hivebrite.com>

### Description

As a Terraform user in some cases I need to read endpoints objects.
An example is: I need to set a network policy on an workload and this one need to request the Kubernetes api. In this case the cluster IP doesn't work I need IPs from the endpoints object

### Acceptance tests
- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch?

Output from acceptance testing:

```
KUBE_CONFIG_PATH="~/.kube/config" make testacc TESTARGS='-run=TestAccKubernetesDataSourceEndpoints_basic'
==> Checking that code complies with gofmt requirements...
go vet .
TF_ACC=1 go test "/home/damien/go/src/github.com/hashicorp/terraform-provider-kubernetes/kubernetes" -v -run=TestAccKubernetesDataSourceEndpoints_basic -timeout 3h
=== RUN   TestAccKubernetesDataSourceEndpoints_basic
--- PASS: TestAccKubernetesDataSourceEndpoints_basic (8.73s)
PASS
ok      github.com/hashicorp/terraform-provider-kubernetes/kubernetes   8.764s
```

### Release Note
Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-kubernetes/blob/main/CHANGELOG.md):

```release-note
  * Add `kubernetes_endpoints_v1` data source
```

### References

https://github.com/hashicorp/terraform-provider-kubernetes/issues/1328

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### CODE REVIEW

**Comment by Dudesons** [comment]
Hi, @BBBmau @jrhouston (I try with someone else)
There is something missing for reviewing this pull request ?

File: kubernetes/data_source_kubernetes_endpoints_test.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+					resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.0.port.0.name", "httptransport"),
+					resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.0.port.0.port", "80"),
+					resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.0.port.0.protocol", "TCP"),
+				),
+			},
+			{
+				Config: testAccKubernetesEndpointsConfig_basic(name) + testAccKubernetesDataSourceEndpointsConfig_read(),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckKubernetesEndpointExists("kubernetes_endpoints.test", &conf),
+					resource.TestCheckResourceAttr("kubernetes_endpoints.test", "metadata.0.name", name),
```
Comment (arybolovlev) [inline kubernetes/data_source_kubernetes_endpoints_test.go:42]: In the second step, we need to validate the data source resource, but for now, we validate the `kubernetes_endpoints` resource.

Please, change the first argument here and below to `data.kubernetes_endpoints_v1.test`.

File: kubernetes/data_source_kubernetes_endpoints_test.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+					resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.0.port.0.port", "80"),
+					resource.TestCheckResourceAttr("kubernetes_endpoints.test", "subset.0.port.0.protocol", "TCP"),
+				),
+			},
+		},
+	})
+}
+
+func testAccKubernetesDataSourceEndpointsConfig_read() string {
+	return fmt.Sprintf(`data "kubernetes_endpoints" "test" {
```
Comment (arybolovlev) [inline kubernetes/data_source_kubernetes_endpoints_test.go:59]: Please, use the data source name `kubernetes_endpoints_v1`.

File: kubernetes/provider.go [hunk 3]
```diff
@@ -213,6 +213,8 @@ func Provider() *schema.Provider {
 			"kubernetes_all_namespaces":             dataSourceKubernetesAllNamespaces(),
 			"kubernetes_secret":                     dataSourceKubernetesSecret(),
 			"kubernetes_secret_v1":                  dataSourceKubernetesSecret(),
+			"kubernetes_endpoints":                  dataSourceKubernetesEndpoint(),
```
Comment (arybolovlev) [inline kubernetes/provider.go:216]: For new resources and data sources, we use names with the API versions only. Please, remove `kubernetes_endpoints` and leave only `kubernetes_endpoints_v1`.

File: kubernetes/data_source_kubernetes_endpoints_test.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		PreCheck:          func() { testAccPreCheck(t) },
+		IDRefreshName:     "kubernetes_endpoints.test",
+		ProviderFactories: testAccProviderFactories,
+		CheckDestroy:      testAccCheckKubernetesEndpointDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccKubernetesEndpointsConfig_basic(name),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckKubernetesEndpointExists("kubernetes_endpoints.test", &conf),
+					resource.TestCheckResourceAttr("kubernetes_endpoints.test", "metadata.0.name", name),
```
Comment (arybolovlev) [inline kubernetes/data_source_kubernetes_endpoints_test.go:26]: Could you please replace `kubernetes_endpoints.test` and `data.kubernetes_endpoints_v1.test` with variables? For example, as in this [test file](https://github.com/hashicorp/terraform-provider-kubernetes/blob/main/kubernetes/data_source_kubernetes_mutating_webhook_configuration_v1_test.go#L13)

File: kubernetes/data_source_kubernetes_endpoints_test.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
+package kubernetes
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	api "k8s.io/api/core/v1"
```
Comment (arybolovlev) [inline kubernetes/data_source_kubernetes_endpoints_test.go:9]: Please replace this import with `corev1 "k8s.io/api/core/v1"`.

File: kubernetes/data_source_kubernetes_endpoints.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package kubernetes
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (arybolovlev) [inline kubernetes/data_source_kubernetes_endpoints.go:8]: Please replace this import with `metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"`.

File: kubernetes/data_source_kubernetes_endpoints.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+)
+
+func dataSourceKubernetesEndpoint() *schema.Resource {
```
Comment (arybolovlev) [inline kubernetes/data_source_kubernetes_endpoints.go:11]: Please add a `V1` suffix to the function name and make it plural to align with `kind` and `apiVersion` -- `dataSourceKubernetesEndpointsV1`.

File: kubernetes/data_source_kubernetes_endpoints.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+				Description: "Set of addresses and ports that comprise a service. More info: https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors",
+				Optional:    true,
+				Elem:        schemaEndpointsSubset(),
+				Set:         hashEndpointsSubset(),
+			},
+		},
+	}
+}
+
+func dataSourceKubernetesEndpointRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```
Comment (arybolovlev) [inline kubernetes/data_source_kubernetes_endpoints.go:27]: Please add a V1 suffix to the function name and make it plural to align with `kind` and `apiVersion` -- `dataSourceKubernetesEndpointsV1Read`.

File: kubernetes/data_source_kubernetes_endpoints_test.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+)
+
+func TestAccKubernetesDataSourceEndpoints_basic(t *testing.T) {
```
Comment (arybolovlev) [inline kubernetes/data_source_kubernetes_endpoints_test.go:12]: Please add a `V1` suffix to the function name to align with `kind` and `apiVersion` -- TestAccKubernetesDataSourceEndpointsV1_basic.

File: kubernetes/data_source_kubernetes_endpoints.go [hunk 10]
```diff
@@ -0,0 +1,35 @@
+package kubernetes
```
Comment (arybolovlev) [inline kubernetes/data_source_kubernetes_endpoints.go:1]: Please rename this file with the `apiVersion` suffix -- `data_source_kubernetes_endpoints_v1.go` and the same for the related test file.

**Comment by arybolovlev** [comment]
For the changelog file, please use the tag `feature`, something like:
````
```release-note:feature
New data source: `kubernetes_endpoints_v1`
```
````

Looking forward to your changes!

**Comment by Dudesons** [comment]
Hi, @arybolovlev ,

Thank you for your review, I did changes you're requested.
There are 2 parts where I still calling the `kubernetes_endpoints` because the function I'm using from the resource is still with the "old style":
 * https://github.com/hashicorp/terraform-provider-kubernetes/pull/1805/files#diff-f9b10f8d821475b2d654788aa63d2e14e484cd3ead3dee77aff8ac99059cdcebR15
 * https://github.com/hashicorp/terraform-provider-kubernetes/pull/1805/files#diff-f9b10f8d821475b2d654788aa63d2e14e484cd3ead3dee77aff8ac99059cdcebR63

I have a question, in the long term vision this provider wants to use resources with `kind` and `apiVersion` instead only the `kind` ?
I ask that because I'm using this provider since 3years now in production and having only the kind to manage was useful for version migration. Maybe one time we had a strange behavior when we had the hpa v2beta2 but not sure

**Comment by Dudesons** [comment]
I run test again on my local cluster:
```
 KUBE_CONFIG_PATH="~/.kube/config" make testacc TESTARGS='-run=TestAccKubernetesDataSourceEndpointsV1_basic'
==> Checking that code complies with gofmt requirements...
go vet .
TF_ACC=1 go test "/home/damien/go/src/github.com/hashicorp/terraform-provider-kubernetes/kubernetes" -v -run=TestAccKubernetesDataSourceEndpointsV1_basic -timeout 3h
=== RUN   TestAccKubernetesDataSourceEndpointsV1_basic
--- PASS: TestAccKubernetesDataSourceEndpointsV1_basic (6.22s)
PASS
ok      github.com/hashicorp/terraform-provider-kubernetes/kubernetes   6.251s
```

File: kubernetes/data_source_kubernetes_endpoints_v1.go [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func dataSourceKubernetesEndpointV1() *schema.Resource {
```
Comment (arybolovlev) [inline kubernetes/data_source_kubernetes_endpoints_v1.go:11]: Could you please rename `dataSourceKubernetesEndpointV1` to `dataSourceKubernetesEndpointsV1`?

File: website/docs/d/endpoint_v1.html.markdown [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+# kubernetes_endpoints_v1
+
+An Endpoints resource is an abstraction, linked to a Service, which defines the list of endpoints that actually implement the service.
+
+
+## Example Usage
+
+```hcl
+data "kubernetes_endpoints" "api_endpoints" {
```
Comment (arybolovlev) [inline website/docs/d/endpoint_v1.html.markdown:17]: Could you please change data source name from `kubernetes_endpoints` to `kubernetes_endpoints_v1`?

**Review by arybolovlev — COMMENTED** [review]
Hi @Dudesons,

Looks great! Thanks a lot, just a small rename fix in the doc and function name. Once it is done we are good to merge and a new data source will be available in the upcoming release.

Thanks! 🚀👍🏻

**Comment by Dudesons** [comment]
Hi @arybolovlev,

Sorry for the delay, I just did the update

**Review by arybolovlev — APPROVED** [review]
Thank you @Dudesons! 🚀 Great job! 👍🏻

**Pull request merged**